src/utilities/include/svd.h

```diff
@@ -0,0 +1,381 @@
+// MIT License
+
+// Copyright (c) 2019 wi-re
+// Copyright 2023 The Manifold Authors.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// Modified from https://github.com/wi-re/tbtSVD, removing CUDA dependence and
+// approximate inverse square roots.
+
+#include <algorithm>
+#include <cmath>
+#include <tuple>
+#include <utility>
+
+namespace SVD {
+// Constants used for calculation of givens quaternions
+constexpr float _gamma = 5.828427124f;   // sqrt(8)+3;
+constexpr float _cStar = 0.923879532f;   // cos(pi/8)
+constexpr float _sStar = 0.3826834323f;  // sin(pi/8)
+// Threshold value
+constexpr float _SVD_EPSILON = 1e-6f;
+// Iteration counts for Jacobi Eigen Analysis, influence precision
+constexpr int JACOBI_STEPS = 12;
+
+// Helper function used to swap X with Y and Y with  X if c == true
+void condSwap(bool c, float& X, float& Y) {
+  float Z = X;
+  X = c ? Y : X;
+  Y = c ? Z : Y;
+}
+// Helper function used to swap X with Y and Y with -X if c == true
+void condNegSwap(bool c, float& X, float& Y) {
+  float Z = -X;
+  X = c ? Y : X;
+  Y = c ? Z : Y;
+}
+// Helper class to contain a quaternion. Could be replaced with float4 (CUDA
+// based type) but this might lead to unintended conversions when using the
+// supplied matrices
+struct quaternion {
+  float x = 0.f, y = 0.f, z = 0.f, w = 1.f;
+  float& operator[](int32_t arg) { return ((float*)this)[arg]; }
+};
+// A simple 3x3 Matrix class
+struct Mat3x3 {
```

Comment:
Should really replace this with glm::mat3 for consistency and to remove code. Also, I'm a little concerned I have my row-major/column-major backwards, but it's passing the tests... We should probably have more.

Reply:
I think eigenvalue is invariant under matrix transpose?

Reply:
Oh right, that makes sense.

Reply:
Okay, I verified that we now have the correct transpose for the actual decomposition too.